core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -2502,9 +2502,10 @@ private RelBuilder pruneAggregateInputFieldsAndDeduplicateAggCalls(
           newProjects.add(project.getProjects().get(i));
           builder.add(project.getRowType().getFieldList().get(i));
         }
+
         r =
-            project.copy(cluster.traitSet(), project.getInput(), newProjects,
-                builder.build());
+            project.copy(project.getTraitSet().apply(targetMapping), project.getInput(),
```

Comment:
If that's not the right one, it would result in corrupting the Collation of the underlying Project

Reply:
It looks like it maps to the underlying project to me.
The same fields referenced in building the Mapping are used to construct the newProjects variable (the list of RexNodes in the new projection).

Reply:
The use of the mapping was to handle the collation and distribution traits.

Reply:
I see, then the current test is not complete enough, it must pass with your variant while it must fail if you remove the mapping application, can you rework it to make this happen?

Reply:
I haven't figured out how to create a test that covers this -- do you know of a way to alter collations for nodes created by a RelBuilder?

I'm not entirely sure if using the mapping is necessary. When implementing this using the Mapping route vs. copy the traitset, I see a slight difference in the interner, but I don't think this would be user-facing:
![image](https://github.com/apache/calcite/assets/1657237/f8b955e9-f47b-454e-93e0-d4fd49b1a9c2)
The order of the entries in the MapMakerInternalMap differs a bit (the PRESERVE entry appears earlier).

Reply:
@asolimando , what I've found now:
1. There doesn't appear to be a way to add a RelDistribution trait to the builder and have it apply to the projection. Adding a RelDistribution trait using builder.exchange() adds a LogicalExchange node on in between the mapped projections so the pruning block doesn't get triggered.
2. It is possible to use builder.sort() to add a RelCollation trait to the projection before the mapping. However the RelCollation is stored in a RelCompositeTrait, and that composite trait does not override apply(Mapping), it just uses the default implementation which is the identity mapping. This seems off though -- I don't get why it doesn't propagate to wrapped traits. I don't have enough background to tell if this is intentional or a bug. What are your thoughts here?

Reply:
Hey @asolimando ,

I've changed the implementation to skip the mapping and removed the test for other traits as I had issues with them described above. Perhaps it'd be better to deal with other traits in a separate ticket.

Reply:
Can you file a ticket for the issue you encountered with composite traits and missing override of mapping?

You can add in the description the unit tests for collations, link this ticket and say that the mapping in such case would be needed but it is ineffective.

Regarding the current PR, once you have the other ticket, I'd just add a comment pointing to it where you have the `project.copy` part so to make sure that this is a known limitation for people down the line.

Reply:
I've created https://issues.apache.org/jira/browse/CALCITE-6391 for the propagation of apply() calls in RelCompositeTrait. Updating the code to reference this.

Reply:
Done.